src/endpoints/sources.js

```diff
@@ -105,10 +131,45 @@ const canonicalizeDataset = (pathBuilder) => (req, res, next) => {
  */
 const ifDatasetExists = async (req, res, next) => {
   if (!(await req.context.dataset.exists())) throw new NotFound();
-  next();
+  return next();
 };
 
 
+/* XXX TODO: Support automatically translating v1 (meta and tree) to v2
+ * (main) if the latter is requested but only the former is available?
+ * We could, but maybe not worth it for these new endpoints, unless/until we
+ * want to point Auspice at these endpoints and still support v1.
+ *   -trs, 22 Nov 2021
+ */
+const sendDatasetSubresource = type =>
+  sendSubresource(req => req.context.dataset.subresource(type));
+
+
+const getDatasetSubresource = type => contentTypesProvided([
```

Comment:
I found this section hard to get my head around, and I think it'll be tricky to debug when something goes wrong!

We're sometimes calling the (higher order function) `contentTypesProvided` in a nested fashion,
1. once to set up a iterator across the provided handlers for a content type (this is `getDataset`)
2. once again across a provided handler to set up an iterator across different content-types (`getDatasetSubresource`)

I must be missing something important here -- why can we not get rid of these helper function and have the following?

```js
const getDataset = contentTypesProvided([
  ["text/html", ifDatasetExists, sendAuspiceEntrypoint],
  ["application/json", sendDatasetSubresource("main")],
  ["application/vnd.nextstrain.dataset.main+json", sendDatasetSubresource("main")],
  ["application/vnd.nextstrain.dataset.root-sequence+json", sendDatasetSubresource("root-sequence")],
  ["application/vnd.nextstrain.dataset.tip-frequencies+json", sendDatasetSubresource("tip-frequencies")],
```

Reply:
Ah, good question! You're not missing anything! The simplification you give as an example would work just fine here in this PR, and quite possibly is what I should have committed for this PR.

The extra `getDatasetMain`, `getDatasetRootSequence`, etc. handlers (with their own content-type dispatch) exist only because I anticipated using them to handle the convenience routes discussed in the [partial design doc](https://docs.google.com/document/d/1HzJidWHb4GaHtigXmqD8NF8VrnQm-oLry-lBQZx-D7I/edit)), e.g. `/ncov/open/global.json`, `/ncov/open/global_root-sequence.json`, etc. I actually have a prototype of those routes which I omitted from this PR because they weren't complete yet and I wanted to get the rest of the work out for review sooner than later.

Reply:
Ok, I don't fully grasp how exactly those convenience routes would work without seeing code but I do worry about the complexity we are introducing here. A "linear" stack of express middleware is easy to reason about (at least, once one is familiar with the behavior of express!) but we now have a stack of middleware where some layers iterate over a set of (async) handlers¹ and some of those handlers also iterate over their own set of (async) handlers (and maybe those handlers will iterate over their own handlers etc etc).

¹ and these handlers have arguments that make them look like express middleware, which I think is by design so they can be used as express middleware in some circumstances? (I can't find an example of this, but there probably is one.)

Reply:
The partial prototype is in 9592f5c, if you want to 👀.

I hear your concern about complexity, but it's not one I share here. It seems to me that once you understand Express' handler stacks, understanding nested handler stacks isn't much a leap. (And Express itself nests stacks too.)

There are certainly downsides to using declarative handler stacks for routes instead of only using one imperative handler per route, but that's a bigger topic which goes against Express' design and which I think falls outside the scope of this PR if we wanted to consider switching paradigms.

> which I think is by design so they can be used as express middleware in some circumstances?

Yes, by design, not so that we can necessarily use them as Express middleware, but so that our stacks of handlers are consistent (as possible) regardless of if they're called directly or indirectly by Express.